### DISCUSSION TITLE
How to serialize a 3rd party container?

### DISCUSSION THREAD

**Discussion by magras** [jam1garner/binrw#300] [Q&A] [answered]
There is a way to deserialize a 3rd party container like that:
```rust
#[binread]
struct S {
  #[br(temp)]
  len: u8,
  #[br(parse_with = binrw::helpers::count(len as usize))]
  container: arrayvec::ArrayVec<u8, 255>,
}
```

But how to serialize this container? Is it possible only with custom `#[writer]`?

**Comment by csnover** [comment, marked as answer]
Hi, and thanks for your question! If the third-party container can be converted to a slice, then it can be written by e.g. `#[bw(map = |c| &c[..])]`, which will then use binrw’s built-in write implementation for slices.

**Reply by magras** [reply]
@csnover, thank you, it works perfectly with `arrayvec`.
For others, I find `#[bw(map = AsRef::as_ref)]` to be slightly more readable, but it's a matter of personal preference.

While I don't need it at the moment, I'm still curious if there's an attribute that can take `IntoIterator` or similar trait for non-contiguous containers like `VecDeque`.

**Reply by csnover** [reply]
> I'm still curious if there's an attribute that can take `IntoIterator` or similar trait for non-contiguous containers like `VecDeque`.

Right now there isn’t anything within binrw like that, but only for historical reasons (binrw started life as a read-only library). It’s definitely something that is desired, though! It just requires a community contribution. I think right now there is only a tracking ticket for the specific issue of handling write-indirection, at #4. It would be fine to open an enhancement request, or, author some write-versions of the existing helpers and submit a PR. :-)

**Reply by magras** [reply]
I'm not entirely sure if I'm qualified to write library code in rust. I've had some troubles without generic specialization when dealing with conflicts between `IntoIterator` and concrete container types. However, if I manage to develop a working solution, I'd be happy to contribute it.
